### ISSUE TITLE
GraphSage Link Prediction: Sampling Data

### ISSUE THREAD

**Issue by rmwheeless** [amaiya/ktrain#208]
What is the significance of the sampled edges in the Link Prediction Ipynb Jupyter Notebook under Overview (looks like **link prediction with graph neural networks (GraphSAGE) [example notebook]**)?

** Sampled 527 positive and 527 negative edges. **
** Sampled 475 positive and 475 negative edges. **

I get that it is an attempt to deal with oversized data sets but I do not get exactly how it works.  The first number seems to correspond to the Original Graph and the Second to the Validation Graph/Set.  Could someone please explain to me exactly how this process works?

Also, how might one display a graph of both the Positive and Negative Edges; are both types of edges already inherent in the Validation Graph?

Thanks so much!

**Comment by amaiya** [comment]
Think of the sampled edges as node pairs.  The positive node pairs have links in the original graph.  The negative node pairs don't.  This is the training set, which is no different than a labeled sample of positive movie reviews and negative movie reviews for text classification (or a labeled sample of cat and dog pictures for image classification).   The positive and negative node pairs are sampled so that they are the same number (balanced training set).  

Given a graph with the sampled edges removed and these node pairs, the model will learn to accurately predict when a link should exist between a pair of nodes and when one should not.    A set of positive and negative node pairs are also sampled for validation in a similar fashion and is used to evaluate predictions after training.  So, we take 10% edges for validation and sample another 10% of the remaining edges for training (which is why the training sample is smaller than validation sample.  The graph used to train the model has both the training and validation edges removed: `4276 edges = 5278 total edges - 527 validation edges - 475 training edges`.  

As shown in the notebook, the graphs themselves are [networkx](https://networkx.github.io/) graphs, so you can use that library to visualize them (or export the node/edge list and visualize in various other tools like [Gephi](https://gephi.org/) or [Cytoscape](https://cytoscape.org/)). You'll have to refer to the documentation of those libraries for more information on how to do it.  The [igraph](https://igraph.org/) library is also highly recommended.

**Comment by rmwheeless** [comment]
Arun,

I am really really sorry to bother you yet again, but if one displayed the graph of the Validation Set, would this show both the negative and positive edges aka links?

Thanks,
R. Maxwell Wheeless

________________________________

**Comment by amaiya** [comment]
>  The graph used to train the model has both the training and validation edges removed: 4276 edges = 5278 total edges - 527 validation edges - 475 training edges.

As mentioned above, the training graph has both training and validation edges **removed**.   The validation graph has just the validation edges removed.  This is demonstrated in the notebook.  If you want to access the removed training node pairs, for example, they are accessible with: `trn.link_seq.ids`.  The labels (positive or negative) should be accessible with: `trn.link_seq.targets`.
